src/common/cid.ts

```diff
@@ -10,17 +18,26 @@ export function decodeCID(val: CID | object | string): CID {
   const cid = CID.asCID(val)
   if (cid) return cid
 
-  if (typeof val === "string") return CID.parse(val as string)
+  if (typeof val === "string") return CID.parse(val)
   if (typeof val === "object" && "version" in val && "code" in val && "multihash" in val) {
     return CID.create(val.version, val.code, val.multihash)
   }
 
-  throw new Error(`Could not decode CID: ${val}`)
+  // Result of CID.toJSON()
+  if (typeof val === "object" && hasProp(val, "version") && val.version === 1 && hasProp(val, "code") && isNum(val.code) && hasProp(val, "hash") && isObject(val.hash) && Object.values(val.hash).every(isNum)) {
```

Comment:
this condition is a little verbose. I wonder if it would make sense to save it to a variable with a more descriptive name and check against that? Maybe with a comment too? something like:
```suggestion
  /**
   * A valid encoded CID object must meet the following criteria:
   * - has a `version` equal to 1
   * - has a `code` that is a number
   * - has a `hash` object containing only numerical values
  */
  const isCIDValid = typeof val === "object" && hasProp(val, "version") && val.version === 1 && hasProp(val, "code") && isNum(val.code) && hasProp(val, "hash") && isObject(val.hash) && Object.values(val.hash).every(isNum)
  
  if (isCIDValid) {
```

Reply:
Your comment suggestion is not entirely correct. A valid CID can take many shapes, hence the other parts of that function. This is a particular format coming from `multiformats/cid -> instance.toJSON`. I'll add a better explanation 👍

Reply:
good to know, thanks! I was just trying to distill the condition down into a comment 😅